core/src/poh_recorder.rs

```diff
@@ -51,6 +51,80 @@ type Result<T> = std::result::Result<T, PohRecorderError>;
 pub type WorkingBankEntry = (Arc<Bank>, (Entry, u64));
 pub type BankStart = (Arc<Bank>, Arc<Instant>);
 
+pub struct Record {
+    pub mixin: Hash,
+    pub transactions: Vec<Transaction>,
+    pub slot: Slot,
+    pub sender: Sender<Result<()>>,
+}
+impl Record {
+    pub fn new(
+        mixin: Hash,
+        transactions: Vec<Transaction>,
+        slot: Slot,
+        sender: Sender<Result<()>>,
+    ) -> Self {
+        Self {
+            mixin,
+            transactions,
+            slot,
+            sender,
+        }
+    }
+}
+
+pub struct Recorder {
+    // shared by all users of PohRecorder
+    pub record_sender: Sender<Record>,
+    // unique to this caller
+    pub result_sender: Sender<Result<()>>,
+    pub result_receiver: Receiver<Result<()>>,
+}
+
+impl Clone for Recorder {
+    fn clone(&self) -> Self {
+        Recorder::new(self.record_sender.clone())
+    }
+}
+
+impl Recorder {
+    pub fn new(record_sender: Sender<Record>) -> Self {
+        let (result_sender, result_receiver) = channel();
+        Recorder {
+            // shared
+            record_sender,
+            // unique to this caller
+            result_sender,
+            result_receiver,
+        }
+    }
+    pub fn record(
+        &self,
+        bank_slot: Slot,
+        mixin: Hash,
+        transactions: Vec<Transaction>,
+    ) -> Result<()> {
+        let res = self.record_sender.send(Record::new(
+            mixin,
+            transactions,
+            bank_slot,
+            self.result_sender.clone(),
+        ));
+        if res.is_err() {
+            return Err(PohRecorderError::MaxHeightReached); // TODO what error here
+        }
+        let res = self
+            .result_receiver
+            .recv_timeout(std::time::Duration::from_millis(2000)); //TODO: consider timeout? _timeout(Duration::from_millis(4000));
+        match res {
+            Err(_err) => {
+                Err(PohRecorderError::MaxHeightReached) // TODO - what error here
```

Comment:
@sakridge and I discussed and it looks like MaxHeightReached is the correct behavior in all callers. This should only occur in real error conditions or in shutdown of tests. We could add a new error code for PohRecorderError indicating a different kind of failure but it looks like we want the same behavior as MaxHeightReached, presumably so caller tries again.